### ISSUE TITLE
Cronjob blocking update (forever)

### ISSUE THREAD

**Issue by jancborchardt** [nextcloud/server#9992]
Both @skjnldsv and me had this issue earlier, when an update gave us this message:

> Waiting for cron to finish (check again in 5 seconds)

And that just kept coming again and again. Setting maintenance mode to false also does not work.

It can be fixed using `UPDATE oc_jobs SET reserved_at=0;` (thanks @nickvergessen), but should probably not happen.

cc @MorrisJobke @nickvergessen

**Comment by MorrisJobke** [comment]
> It can be fixed using UPDATE oc_jobs SET reserved_at=0 (thanks @nickvergessen), but should probably not happen.

This was added on purpose to not have a running background job while updating. This is not really a regression, but a symptom of a previously broken behavior that was just ignored. Why do you have jobs that where reserved but never finished? Which jobs are still reserved?

**Comment by MorrisJobke** [comment]
ref #9900

**Comment by jancborchardt** [comment]
> Why do you have jobs that where reserved but never finished? Which jobs are still reserved?

I have no idea – it’s on my local dev instance, and these were the ones with a reserved_at value of over 0:
```
INSERT INTO oc_jobs VALUES(5,'OCA\DAV\CardDAV\SyncJob','null',1515071551,1529926705,1529926705,0);
INSERT INTO oc_jobs VALUES(30,'OCA\Deck\Cron\DeleteCron','null',1517851983,1529926705,1529926705,0);
```

**Comment by MorrisJobke** [comment]
Anything in the logs which says "cron" as app?

**Comment by jancborchardt** [comment]
@skjnldsv did you have anything like that?

I need to check back tomorrow, am mobile only.

**Comment by skjnldsv** [comment]
I had an error about a missing index 0 (was on the level array a few lines above the waitforcron function)
Aside from that, nothing! :(

**Comment by MorrisJobke** [comment]
Ping me once you have this again and do some screensharing/pair-debugging

**Comment by skjnldsv** [comment]
@MorrisJobke The pending Job I had was a backgroundTest coming after running a phpunit test iirc
You might be able to recreate it?
Create a fake job entry into the db and run the upgrader

**Comment by MorrisJobke** [comment]
Sure - I can recreate by manipulate the DB but that’s not the point. I want to know why the job was started and never finished.

**Comment by nickvergessen** [comment]
Errored here as well just now:
Reserved jobs are: `OCA\Federation\SyncJob` and `OCA\UpdateNotification\Notification\BackgroundJob`

But neither app is enabled, so when they are tried to be executed, the class does not exist => error => catched => ignored and continued

The cron.php ignores those for 12 hours, to save time and not try them again and again, because that caused a lot of troubles for non system crons in the past.

**Comment by MorrisJobke** [comment]
> But neither app is enabled, so when they are tried to be executed, the class does not exist => error => catched => ignored and continued
> 
> The cron.php ignores those for 12 hours, to save time and not try them again and again, because that caused a lot of troubles for non system crons in the past.

Any idea how to check then for proper cron execution? Wouldn't it help to put the "last checked" to 12 h in the future?

**Comment by MorrisJobke** [comment]
> But neither app is enabled, so when they are tried to be executed, the class does not exist => error => catched => ignored and continued

Weird - this case should be catched by the `buildJob` call and not cause this entry: https://github.com/nextcloud/server/blob/18e9631810ad1d3d72c2b4bbee330169808108ad/lib/private/BackgroundJob/JobList.php#L253-L277

**Comment by MorrisJobke** [comment]
At least I can reproduce it and will now debug it locally.

* disable updatenotification and federation app
* run cron.php
* two jobs that @nickvergessen get the reserved at set to a non-zero value

**Comment by nickvergessen** [comment]
Reservation is done before build:
https://github.com/nextcloud/server/blob/18e9631810ad1d3d72c2b4bbee330169808108ad/lib/private/BackgroundJob/JobList.php#L205-L214

It should also be like this, otherwise two could try this in parallel...

**Comment by MorrisJobke** [comment]
Fix is in #10010

**Comment by jancborchardt** [comment]
For the record, just happened again and the last log entries are:
```
{"reqId":"GgknU1m91yQ3UKhedON6","level":0,"time":"2018-06-26T14:53:52+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/nextcloud\/index.php\/apps\/files\/?dir=\/&fileid=7","message":"Scss is disabled for \/home\/jan\/nextcloud\/core\/css\/jquery-ui-fixes.scss, ignoring","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:60.0) Gecko\/20100101 Firefox\/60.0","version":"14.0.0.6"}
{"reqId":"GgknU1m91yQ3UKhedON6","level":0,"time":"2018-06-26T14:53:52+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/nextcloud\/index.php\/apps\/files\/?dir=\/&fileid=7","message":"Scss is disabled for \/home\/jan\/nextcloud\/core\/css\/server.scss, ignoring","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:60.0) Gecko\/20100101 Firefox\/60.0","version":"14.0.0.6"}
{"reqId":"GgknU1m91yQ3UKhedON6","level":0,"time":"2018-06-26T14:53:52+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/nextcloud\/index.php\/apps\/files\/?dir=\/&fileid=7","message":"Scss is disabled for \/home\/jan\/nextcloud\/core\/css\/css-variables.scss, ignoring","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:60.0) Gecko\/20100101 Firefox\/60.0","version":"14.0.0.6"}
{"reqId":"GgknU1m91yQ3UKhedON6","level":0,"time":"2018-06-26T14:53:52+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/nextcloud\/index.php\/apps\/files\/?dir=\/&fileid=7","message":"Scss is disabled for \/home\/jan\/nextcloud\/core\/css\/jquery.ocdialog.scss, ignoring","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:60.0) Gecko\/20100101 Firefox\/60.0","version":"14.0.0.6"}
{"reqId":"GgknU1m91yQ3UKhedON6","level":0,"time":"2018-06-26T14:53:52+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"GET","url":"\/nextcloud\/index.php\/apps\/files\/?dir=\/&fileid=7","message":"Scss is disabled for \/home\/jan\/nextcloud\/core\/search\/css\/results.scss, ignoring","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:60.0) Gecko\/20100101 Firefox\/60.0","version":"14.0.0.6"}
{"reqId":"sK7ehz82G1Wo3dl2z2g3","level":1,"time":"2018-06-26T14:53:56+00:00","remoteAddr":"127.0.0.1","user":"--","app":"updater","method":"GET","url":"\/nextcloud\/core\/ajax\/update.php?requesttoken=82VGFUSvHlA5PqudkC917AFySIITHr7nQ%2BGbr0xtr38%3D%3Aqi4BfCfEaH98SMDH%2BUUGgWg6G85SbeSVKqqwmCsLhDY%3D","message":"\\OC\\Updater::setDebugLogLevel: Set log level to debug","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:60.0) Gecko\/20100101 Firefox\/60.0","version":"14.0.0.6"}
{"reqId":"sK7ehz82G1Wo3dl2z2g3","level":1,"time":"2018-06-26T14:53:56+00:00","remoteAddr":"127.0.0.1","user":"--","app":"updater","method":"GET","url":"\/nextcloud\/core\/ajax\/update.php?requesttoken=82VGFUSvHlA5PqudkC917AFySIITHr7nQ%2BGbr0xtr38%3D%3Aqi4BfCfEaH98SMDH%2BUUGgWg6G85SbeSVKqqwmCsLhDY%3D","message":"\\OC\\Updater::maintenanceEnabled: Turned on maintenance mode","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:60.0) Gecko\/20100101 Firefox\/60.0","version":"14.0.0.6"}
```

Will test the PR fix :)

**Comment by juliushaertl** [comment]
This still happens to me on the latest master.

This seems to be OCA\DAV\BackgroundJob\UpdateCalendarResourcesRoomsBackgroundJob for me:
```
sqlite> select * from oc_jobs where reserved_at > 0;
5|OCA\DAV\BackgroundJob\UpdateCalendarResourcesRoomsBackgroundJob|null|1531300751|1531300751|1531300751|0
```

@MorrisJobke

**Comment by MorrisJobke** [comment]
1531300751 this is 3,5h ago ... maybe the cron job was interrupted? What says your log?

**Comment by MorrisJobke** [comment]
> This still happens to me on the latest master.

Are you sure it was on latest master and not on an old branch of a PR?

**Comment by nickvergessen** [comment]
The thing is, if you where on an old branch sometime in between, that could have bricked the database again.

**Comment by MorrisJobke** [comment]
> The thing is, if you where on an old branch sometime in between, that could have bricked the database again.

Ah ... yes, that's true :/

**Comment by juliushaertl** [comment]
According to the log it seems to be related to the UpdateCalendarResourcesRoomsBackgroundJob which throws a TypeError if no backend for the resource booking is available. 

https://github.com/nextcloud/server/issues/10203

**Comment by morph027** [comment]
Just wanted to upgrade from beta4 to RC1 and have the same issue again....

**Comment by MorrisJobke** [comment]
> Just wanted to upgrade from beta4 to RC1 and have the same issue again....

Which cron job wasn't finished? and did cron still run?

**Comment by morph027** [comment]
Not sure. How to find out? The cron job itself was'nt running....the log even told me it was unable to run because nextcloud is in maintenance mode when  i was trying to run it manually.

**Comment by sphrak** [comment]
I am having the same problem. I tried both via `./occ upgrade` and the web gui. In a naive attempt to bypass it I tried to set the background job to `ajax` instead but that didnt do anything differently.

**Comment by morph027** [comment]
Mine has a zero in `OCA\DAV\CardDAV\Sync\SyncJob`

```
+------+---------------------------------------------------+-------------------------------------------------------+------------+--------------+-------------+--------------------+
| id   | class                                             | argument                                              | last_run   | last_checked | reserved_at | execution_duration |
+------+---------------------------------------------------+-------------------------------------------------------+------------+--------------+-------------+--------------------+
|    1 | OC\BackgroundJob\Legacy\RegularJob                | ["\\OC\\Files\\Cache\\BackgroundWatcher","checkNext"] | 1501516801 |   1501516802 |           0 |                  0 |
|   11 | OC\BackgroundJob\Legacy\RegularJob                | ["OCA\\News\\Backgroundjob\\Task","run"]              | 1501516801 |   1501516801 |           0 |                  0 |
| 2446 | OCA\Activity\BackgroundJob\ExpireActivities       | null                                                  | 1501459201 |   1501516801 |           0 |                  0 |
| 3071 | OC\BackgroundJob\Legacy\RegularJob                | ["OCA\\News\\Cron\\Updater","run"]                    | 1501516802 |   1501516802 |           0 |                  0 |
| 3131 | OCA\Activity\BackgroundJob\EmailNotification      | null                                                  | 1501516801 |   1501516801 |           0 |                  0 |
| 3135 | OCA\Files_Trashbin\BackgroundJob\ExpireTrash      | null                                                  | 1501516801 |   1501516801 |           0 |                  0 |
| 3136 | OCA\Files_Versions\BackgroundJob\ExpireVersions   | null                                                  | 1501516801 |   1501516801 |           0 |                  0 |
| 3208 | OCA\Files\BackgroundJob\CleanupFileLocks          | null                                                  | 1501516801 |   1501516801 |           0 |                  0 |
| 3252 | OCA\Files\BackgroundJob\ScanFiles                 | null                                                  | 1501516801 |   1501516801 |           0 |                  1 |
| 3253 | OCA\Files\BackgroundJob\DeleteOrphanedItems       | null                                                  | 1501516802 |   1501516802 |           0 |                  0 |
| 3255 | OCA\Federation\SyncJob                            | null                                                  | 1501459202 |   1501516802 |           0 |                  0 |
| 3256 | OCA\DAV\CardDAV\SyncJob                           | null                                                  | 1501459202 |   1501516802 |           0 |                  1 |
| 3257 | OC\BackgroundJob\Legacy\RegularJob                | ["OCA\\Music\\Backgroundjob\\CleanUp","run"]          | 1501516802 |   1501516802 |           0 |                  0 |
| 5814 | \OC\Authentication\Token\DefaultTokenCleanupJob   | null                                                  | 1501516802 |   1501516802 |           0 |                  0 |
| 5816 | OCA\UpdateNotification\Notification\BackgroundJob | null                                                  | 1501459203 |   1501516802 |           0 |                  4 |
| 5818 | OCA\DAV\CardDAV\Sync\SyncJob                      | null                                                  |          0 |   1501516801 |           0 |                  0 |
| 5819 | OCA\Files_Sharing\DeleteOrphanedSharesJob         | null                                                  | 1501516802 |   1501516802 |           0 |                  0 |
| 5820 | OCA\Files_Sharing\ExpireSharesJob                 | null                                                  | 1501459207 |   1501516802 |           0 |                  0 |
| 6834 | OCA\NextcloudAnnouncements\Cron\Crawler           | null                                                  | 1501459201 |   1501516801 |           0 |                  1 |
| 6837 | OCA\Survey_Client\BackgroundJobs\MonthlyReport    | null                                                  | 1500314401 |   1501516802 |           0 |                  3 |
| 7059 | OCA\UpdateNotification\ResetTokenBackgroundJob    | null                                                  | 1501516801 |   1501516801 |           0 |                  0 |
| 7096 | OCA\Deck\Cron\DeleteCron                          | null                                                  | 1501516802 |   1501516802 |           0 |                  0 |
+------+---------------------------------------------------+-------------------------------------------------------+------------+--------------+-------------+--------------------+
```

**Comment by MorrisJobke** [comment]
It checks only if "reserved_at" is set to a number that is bigger than "CURRENT UNIX TIMESTAMP" - 43200 (reserved within the past 12h).

So if the reserved_at is 0 all should be fine. What DB are you using?

**Comment by MorrisJobke** [comment]
You could press "Ctrl +C" and start it again. Because I guess SQLite could cause some trouble here :/ For other DBs this is not a problem usually.

**Comment by morph027** [comment]
Argh, actually i'm using Postgres and the output above was from the "old" MySQL backup....

This is the real one:

```
nextcloud=# select * from oc_jobs;
  id   |                              class                              |                       argument                        |  last_run  | last_checked | reserved_at | execution_duration 
-------+-----------------------------------------------------------------+-------------------------------------------------------+------------+--------------+-------------+--------------------
 11148 | OC\Log\Rotate                                                   | null                                                  | 1535272780 |   1535272786 |           0 |                  0
 14899 | OCA\Spreed\BackgroundJob\ExpireSignalingMessage                 | null                                                  | 1535189900 |   1535278015 |           0 |                  0
    11 | OC\BackgroundJob\Legacy\RegularJob                              | ["OCA\\News\\Backgroundjob\\Task","run"]              | 1535272786 |   1535272786 |           0 |                  0
  7059 | OCA\UpdateNotification\ResetTokenBackgroundJob                  | null                                                  | 1535272786 |   1535272786 |           0 |                  0
  5816 | OCA\UpdateNotification\Notification\BackgroundJob               | null                                                  | 1535266313 |   1535272786 |           0 |                  1
 14756 | OCA\News\Cron\Updater                                           | null                                                  | 1535272780 |   1535272780 |           0 |                  5
  6837 | OCA\Survey_Client\BackgroundJobs\MonthlyReport                  | null                                                  | 1534190960 |   1535272785 |           0 |                  1
 11147 | OC\Authentication\Token\DefaultTokenCleanupJob                  | null                                                  | 1535272786 |   1535272786 |           0 |                  0
  3253 | OCA\Files\BackgroundJob\DeleteOrphanedItems                     | null                                                  | 1535270938 |   1535272785 |           0 |                  0
  5820 | OCA\Files_Sharing\ExpireSharesJob                               | null                                                  | 1535270003 |   1535272786 |           0 |                  0
 14247 | OCA\FullTextSearch\Cron\Index                                   | null                                                  | 1535272785 |   1535272785 |           0 |                  1
 19711 | OCA\Support\BackgroundJobs\CheckSubscription                    | null                                                  | 1535272786 |   1535272786 |           0 |                  0
  3136 | OCA\Files_Versions\BackgroundJob\ExpireVersions                 | null                                                  | 1535271859 |   1535272786 |           0 |                  0
 19706 | OCA\DAV\BackgroundJob\CleanupDirectLinksJob                     | null                                                  | 1535215399 |   1535272786 |           0 |                  0
  3255 | OCA\Federation\SyncJob                                          | null                                                  | 1535268138 |   1535272786 |           0 |                  0
  3131 | OCA\Activity\BackgroundJob\EmailNotification                    | null                                                  | 1535272786 |   1535272786 |           0 |                  0
 19705 | OC\Preview\BackgroundCleanupJob                                 | null                                                  | 1535270940 |   1535272786 |           0 |                  0
  3135 | OCA\Files_Trashbin\BackgroundJob\ExpireTrash                    | null                                                  | 1535271859 |   1535272786 |           0 |                  0
 19708 | OCA\DAV\BackgroundJob\CleanupInvitationTokenJob                 | null                                                  | 1535215400 |   1535272786 |           0 |                  0
  3252 | OCA\Files\BackgroundJob\ScanFiles                               | null                                                  | 1535272786 |   1535272786 |           0 |                  0
 10335 | OCA\Passman\BackgroundJob\ExpireCredentials                     | null                                                  | 1533827303 |   1535305826 |           0 |                  0
 19710 | OCA\AdminAudit\BackgroundJobs\Rotate                            | null                                                  | 1535272786 |   1535272786 |           0 |                  0
 19707 | OCA\DAV\BackgroundJob\UpdateCalendarResourcesRoomsBackgroundJob | null                                                  | 1535270938 |   1535272786 |           0 |                  0
  5818 | OCA\DAV\CardDAV\Sync\SyncJob                                    | null                                                  |          0 |   1535305826 |           0 |                  0
  6834 | OCA\NextcloudAnnouncements\Cron\Crawler                         | null                                                  | 1535268139 |   1535272786 |           0 |                  1
  7096 | OCA\Deck\Cron\DeleteCron                                        | null                                                  | 1535272786 |   1535272786 |           0 |                  0
 19709 | OCA\Files_Sharing\BackgroundJob\FederatedSharesDiscoverJob      | null                                                  | 1535215399 |   1535272786 |           0 |                  0
  3208 | OCA\Files\BackgroundJob\CleanupFileLocks                        | null                                                  | 1535272786 |   1535272786 |           0 |                  0
  2446 | OCA\Activity\BackgroundJob\ExpireActivities                     | null                                                  | 1535270003 |   1535272786 |           0 |                  0
  3071 | OC\BackgroundJob\Legacy\RegularJob                              | ["OCA\\News\\Cron\\Updater","run"]                    | 1535272786 |   1535272786 |           0 |                  0
 15886 | OCA\Deck\Cron\ScheduledNotifications                            | null                                                  | 1535272786 |   1535272786 |           0 |                  0
  5819 | OCA\Files_Sharing\DeleteOrphanedSharesJob                       | null                                                  | 1535272786 |   1535272786 |           0 |                  0
 17987 | OCA\Spreed\BackgroundJob\RemoveEmptyRooms                       | null                                                  | 1535189900 |   1535278015 |           0 |                  0
  3256 | OCA\DAV\CardDAV\SyncJob                                         | null                                                  | 1515316028 |   1535305826 |           0 |                  4
 15576 | OCA\User_LDAP\Jobs\CleanUp                                      | null                                                  | 1520174331 |   1535305826 |           0 |                  0
 15577 | OCA\User_LDAP\Jobs\Sync                                         | null                                                  | 1520174331 |   1535305826 |           0 |                  0
 15575 | OCA\User_LDAP\Jobs\UpdateGroups                                 | null                                                  | 1520174331 |   1535305826 |           0 |                  0
 17988 | OCA\Spreed\BackgroundJob\ResetInCallFlags                       | null                                                  | 1532454631 |   1535305826 |           0 |                  0
  5814 | \OC\Authentication\Token\DefaultTokenCleanupJob                 | null                                                  | 1535272786 |   1535272786 |           0 |                  0
  3257 | OC\BackgroundJob\Legacy\RegularJob                              | ["OCA\\Music\\Backgroundjob\\CleanUp","run"]          | 1535272786 |   1535272786 |           0 |                  0
     1 | OC\BackgroundJob\Legacy\RegularJob                              | ["\\OC\\Files\\Cache\\BackgroundWatcher","checkNext"] | 1535272786 |   1535272786 |           0 |                  0
(41 rows)
```

**Comment by morph027** [comment]
Tried again, whatever it was, it's gone and the updater went through without any further issues.

**Comment by bzdega** [comment]
hey guys,

just updating to new production release 14.0.3 and after the update, going back to Nextcoud-Startpage it says: 

> Warte auf das Beenden von Cron (neue Prüfung in 5 Sekunden)…

waiting for cron to finish...

**Comment by MorrisJobke** [comment]
https://github.com/nextcloud/server/issues/10949#issuecomment-426033160

**Comment by bzdega** [comment]
thank, this works.

**Comment by jancborchardt** [comment]
Btw @MorrisJobke I keep running into this issue. The command in the OP `UPDATE oc_jobs SET reserved_at=0;` still works and immediately after doing that, the update runs through.

Are you saying that next time I run into it I should run `SELECT * FROM oc_jobs WHERE reserved_at <> 0;` to debug where it’s from?

**Comment by MorrisJobke** [comment]
> Are you saying that next time I run into it I should run `SELECT * FROM oc_jobs WHERE reserved_at <> 0;` to debug where it’s from?

Correct.



> Btw @MorrisJobke I keep running into this issue. The command in the OP `UPDATE oc_jobs SET reserved_at=0;` still works and immediately after doing that, the update runs through.

This basically not good, because it means that one background job crashed or it is actually still running. I guess we should remove the check again. I just thought it may be a good idea to do this. :(

**Comment by MorrisJobke** [comment]
Fix in #12188
